### ISSUE TITLE
Incorrect syntax highlighting

### ISSUE THREAD

**Issue by theonenotzero** [Pure-D/code-d#410]
Here is a small sample d file which illustrates the issues.

```d
fn!T();

cast(void function())x;

obj.method();
```

Screenshot from VSCode:
![image](https://user-images.githubusercontent.com/103496370/170051355-92696975-36af-48d1-9a67-2068902cf447.png)

In the first line, `fn` should be highlighted as a function, as it is indeed a template function being called. Also, `T` should be highlighted either as a type (purple in my theme) or have no highlighting at all.

In the second line, the wrong closing parenthesis is matched with the opening parenthesis for the cast operation.

In the third line, the `method` should be highlighted as a function while `obj.` should not.

**Comment by WebFreak001** [comment]
for 1 and 3 it's not possible to determine whether things are types, functions or variables from the syntax, semantic highlighting is needed there, which is simply not yet implemented. (See #299)

Also can't reproduce the issue with your sample code:
![sample code with working syntax highlighting](https://wfr.moe/f6T9kd.png)

**Comment by theonenotzero** [comment]
I don't believe any of my suggestions would require semantic analysis at all, since `!` and `.` are not characters that can appear in an identifier. For the first one, when you find a site that looks like a function call, is might be possible to attempt to backtrack until the last `!` token, which would correctly identify the function name. Similarly, for the third case, by splitting tokens along the `.` character you can easily determine the boundary.

I don't know how this works under the hood though, so I might be missing something.

**Comment by WebFreak001** [comment]
```d
struct someType {}
int someFunction() { return 0; }
int someVariable;
int ufcsFun(int x) { return x; }
void myTemplate(alias symbol)() {}

void main() {
    int x = someFunction;
    int y = someVariable;
    int z = someFunction.ufcsFun;

    myTemplate!x;
    myTemplate!(ufcsFun)();
    myTemplate!someVariable();
    myTemplate!someType;
}
```

all valid and without knowing the definition it's impossible to guess at the function calling site what kind each symbol is.

**Comment by theonenotzero** [comment]
Okay, maybe I was being unclear with with what I was asking for. I understand that it is not feasible to know the *type* of a particular token. However, if a token is being called like a function using (), it is probably a function and that assumption should be made. This is already done. However, I was asking that a function called via !() should also be highlighted like a function, *regardless of what the type actually is.* Right now, the highlighting is 100% wrong, as the template paramater is being highlighted as the function being called.

For your example, `myTemplate` is being called using !() so it should be highlighted as a function since it appears to be one. Additionally, none of `x`, `someVariable`, or `someType` should be highlighted as a function, and a little more parsing can confidently determine that they are not in fact functions.
